akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -0,0 +1,643 @@
+          }
+        }
+        override def postStop(): Unit = {
+          // Unlike in the case of preStart, we don't care about the Hub no longer looking at the queue.
+          if (!logic.isShuttingDown) logic.enqueue(Deregister(id))
+        }
+
+        override def onPush(): Unit = {
+          logic.enqueue(Element(id, grab(in)))
+          if (demand > 0) pullWithDemand()
+        }
+
+        private def pullWithDemand(): Unit = {
+          demand -= 1
+          pull(in)
+        }
+
+        // Make some noise
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          throw new MergeHub.ProducerFailed("Upstream producer failed with exception, " +
+            "removing from MergeHub now", ex)
+        }
+
+        private def onDemand(moreDemand: Long): Unit = {
+          if (moreDemand == MergeHub.Cancel) completeStage()
+          else {
+            demand += moreDemand
+            if (!hasBeenPulled(in)) pullWithDemand()
+          }
+        }
+
+        setHandler(in, this)
+      }
+
+    }
+
+    (logic, Sink.fromGraph(sink))
+  }
+}
+
+/**
+ * A BroadcastHub is a special streaming hub that is able to broadcast streamed elements to a dynamic set of consumers.
+ * It consissts of two parts, a [[Sink]] and a [[Source]]. The [[Sink]] broadcasts elements from a producer to the
+ * actually live consumers it has. Once the producer has been materialized, the [[Sink]] it feeds into returns a
+ * materialized value which is the corresponding [[Source]]. This [[Source]] can be materialized arbitrary many times,
+ * where weach of the new materializations will receive their elements from the original [[Sink]].
+ */
+object BroadcastHub {
+
+  /**
+   * Creates a [[Sink]] that receives elements from its upstream producer and broadcasts them to a dynamic set
+   * of consumers. After the [[Sink]] returned by this method is materialized, it returns a [[Source]] as materialized
+   * value. This [[Source]] can be materialized arbitrary many times and each materialization will receive the
+   * broadcast elements form the ofiginal [[Sink]].
+   *
+   * Every new materialization of the [[Sink]] results in a new, independent hub, which materializes to its own
+   * [[Source]] for consuming the [[Sink]] of that materialization.
+   *
+   * If the original [[Sink]] is failed, then the failure is immediately propagated to all of its materialized
+   * [[Source]]s (possibly jumping over already buffered elements). If the original [[Sink]] is completed, then
+   * all corresponding [[Source]]s are completed. Both failure and normal completion is "remembered" and later
+   * materializations of the [[Source]] will see the same (failure or completion) state. [[Source]]s that are
+   * cancelled are simply removed from the dynamic set of consumers.
+   *
+   * @param clazz Type of elements this hub emits and consumes
+   * @param bufferSize Buffer size used by the producer. Gives an upper bound on how "far" from each other two
```

Comment:
Btw, do you think I should keep the upper limit? I thought it makes sense to stop people from doing stupid things. Removing this limit if it ever becomes necessary (hard to imagine) is binary compatible.